backpack/extensions/module_extension.py

```diff
@@ -12,98 +24,198 @@ class ModuleExtension:
-            Instance of the extension currently running
-        module: torch.nn.Module
-            Instance of the extended module
-        g_inp: [Tensor]
-            Gradient of the loss w.r.t. the inputs
-        g_out: Tensor
-            Gradient of the loss w.r.t. the output
-        bpQuantities:
-            Quantities backpropagated w.r.t. the output
+        self.__params: List[str] = [] if params is None else params
+
+        for param in self.__params:
+            if not hasattr(self, param):
+                raise NotImplementedError(
+                    f"The module extension {self} is missing an implementation "
+                    f"of how to calculate the quantity for {param}. "
+                    f"This should be realized in a function "
+                    f"{param}(extension, module, g_inp, g_out, bpQuantities) -> Any."
+                )
+
+    def backpropagate(
+        self,
+        extension: BackpropExtension,
+        module: Module,
+        g_inp: Tuple[Tensor],
+        g_out: Tuple[Tensor],
+        bpQuantities: Any,
+    ) -> Any:
+        """Backpropagation of additional information through the graph.
+
+        Args:
+            extension: Instance of the extension currently running
+            module: Instance of the extended module
+            g_inp: Gradient of the loss w.r.t. the inputs
+            g_out: Gradient of the loss w.r.t. the output
+            bpQuantities: Quantities backpropagated w.r.t. the output
 
         Returns
-        -------
-        bpQuantities:
             Quantities backpropagated w.r.t. the input
         """
         warnings.warn("Backpropagate has not been overwritten")
 
-    def apply(self, ext, module, g_inp, g_out):
-        """
+    def __call__(
+        self,
+        extension: BackpropExtension,
+        module: Module,
+        g_inp: Tuple[Tensor],
+        g_out: Tuple[Tensor],
+    ) -> None:
+        """Apply all actions required by the extension.
+
         Fetch backpropagated quantities from module output, apply backpropagation
         rule, and attach the result to module input(s).
-        """
-        inp = module.input0
-        out = module.output
 
-        bpQuantities = self.__backproped_quantities(ext, out)
+        Args:
+            extension: current backpropagation extension
+            module: current module
+            g_inp: input gradients
+            g_out: output gradients
+
+        Raises:
+            AssertionError: if there is no saved quantity although extension expects one
+        """
+        delete_old_quantities = not self.__should_retain_backproped_quantities(module)
+        bp_quantity = self.__get_backproped_quantity(
+            extension, module.output, delete_old_quantities
+        )
+        if (
+            extension.expects_backpropagation_quantities()
+            and bp_quantity is None
+            and not is_loss(module)
+            and TORCH_VERSION_AT_LEAST_1_9_0
```

Comment:
Can you explain to me why the `AssertionError` is only raised for specific PyTorch versions?

Reply:
Because only the full_backward_hook executes the modules in exactly the same order as the forward hook.
With the old hook it can happen, that the order is changed by PyTorch. In this case, sometimes the bp_quantitiy is None.

Reply:
Thanks -> resolved.